daemon/response.go

```diff
@@ -281,55 +284,176 @@ func (f fileResponse) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+			// Ignore the return value here as we are breaking out
+			// anyway and not sending more messages.
+			rr.safeSendError(e, err)
 			break
 		}
 
-		writer.WriteByte(0x1E) // RS -- see ascii(7), and RFC7464
-
-		// ignore the error...
-		t, _ := log.Time()
-		if err = enc.Encode(client.Log{
-			Timestamp: t,
-			Message:   log.Message(),
-			SID:       log.SID(),
-			PID:       log.PID(),
-		}); err != nil {
+		if closed := rr.safeSendLog(c, log); closed {
 			break
 		}
+	}
+}
+
+func (rr *journalLineReaderSeqResponse) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json-seq")
+
+	flusher, hasFlusher := w.(http.Flusher)
+	writer := bufio.NewWriter(w)
+	enc := json.NewEncoder(writer)
+
+	// Buffer 128 (arbitrary, seems appropriate) messages, and
+	// the number of readers in errors as we know exactly how
+	// many errors to expect on the channel.
+	c := make(chan systemd.Log, 128)
+	e := make(chan error, len(rr.readers))
+	for _, r := range rr.readers {
+		go rr.logReader(r, c, e)
+	}
+
+	writeError := func(err error) {
+		fmt.Fprintf(writer, `\x1E{"error": %q}\n`, err)
+		logger.Noticef("cannot stream response; problem reading: %v", err)
+	}
+
+	writeLogs := func(logs []systemd.Log) error {
+		// sort by timestamp ascending
+		sort.Slice(logs, func(i, j int) bool {
+			ti, _ := logs[i].Time()
+			tj, _ := logs[j].Time()
+			return ti.Before(tj)
+		})
+
+		for _, l := range logs {
+			writer.WriteByte(0x1E) // RS -- see ascii(7), and RFC7464
+
+			// ignore the error...
+			t, _ := l.Time()
+			if err := enc.Encode(client.Log{
+				Timestamp: t,
+				Message:   l.Message(),
+				SID:       l.SID(),
+				PID:       l.PID(),
+			}); err != nil {
+				return err
+			}
+		}
 
 		if rr.follow {
-			if e := writer.Flush(); e != nil {
-				break
+			if err := writer.Flush(); err != nil {
+				return errCannotWriteToClient
 			}
 			if hasFlusher {
 				flusher.Flush()
 			}
 		}
+		return nil
 	}
-	if err != nil && err != io.EOF {
-		fmt.Fprintf(writer, `\x1E{"error": %q}\n`, err)
-		logger.Noticef("cannot stream response; problem reading: %v", err)
+
+	var logReadersDone int
+	for {
+		var logs []systemd.Log
+
+		// always block read on the first one to ensure we don't waste any
+		// time spinning on a channel that will never have any logs
+		select {
+		case log := <-c:
+			logs = append(logs, log)
+		case err := <-e:
+			log.Println("error waiting for logs:", err)
+			if err != io.EOF {
+				writeError(err)
+			}
+			logReadersDone++
+		}
+
+		// Now we spend a small amount of time batch reading all available logs
+		// on the log/error channel, the reason we do this is to make sure we read
+		// everything available for the initial/final batch of logs (when following) or
+		// we read everything available when not following. It's done like this to
+		// ensure that output is consistent and sorted correctly by timestamp when
+		// multiple log-sources are in play.
```

Comment:
I'm still confused on why this needs to be a different loop: if I understand correctly, we anyway want to exit the function only once all readers have been fetched (or a write error occurs), so I don't understand why we have the timeout and why we don't call `writeLogs()` at the end of everything.

Reply:
I managed to remove the first loop, before adding the timeout I was using a default case, which I hadn't corrected for in the remainer of the code. So the first loop was completly unnecessary. 

One of the challenges about this is if we are in follow-mode, then we never get a end of file from the readers. The readers will continue to exist untill a write-fail occurs (i.e. client disconnects by ctrl-c), where we then need to gracefully but rudely shut down the channel so the readers stop. So we can't just write at end of everything as that will not show any logs to the user when we need to show them in as close to real time as possible. 

The timeout ensures that for non-follow mode that all logs can be read so we can correctly sort them before showing them to the user, and for follow mode it ensures that we don't stay forever in the for loop, but write all the logs we've read in regular intervals, the interval here should be short enough for the user not to notice, but not too short either. I was thinking of maybe using 50 or 100 ms instead of 25, as they seem to be just fine as well.

I tried a solution by just checking if there were no more messages in the go channel by using len(c) == 0, however this then comes down to the size of the channel before we break, sort messages and show them. This results in weird (randomly sorted by each time channel is empty) logs if reading from multiple sources.

Reply:
Mmm... it's still not clear to me. It seems to me that the only conditions which would allow us to exit from the outer `for` loop are `if err := writeLogs(logs); err != nil` and `if logReadersDone == len(rr.readers)`, that is if either the client itself disconnects, or if all the readers are done. There is no case where we would get out of the outer loop because of the timeout, is there?

I can see how the timeout can be useful when we are in follow mode, but it seems to me that it can mess up the logging while we are reading the backlog: suppose that we have a service whose log is huge, and other services who only wrote a couple of lines each, then if we do the grouping and sorting of the timestamps after 25ms, it's likely that we'll have retrieved all the logs belonging to those services who only printed a couple of lines, but just the beginning of the logs of the verbose service; if we sort only what we've got so far and send it to the user, then on the next iteration we will continue reading the log from the verbose service, but these lines might be still older than those (from the other services) which we've sent back to the user.

I'm not sure what the solution could be. We should activate the 25ms timer (for the follow mode) and send the first batch of logs only *after* having read the backlogs. Maybe: before the first iteration, we get the current time; then, once we read a log line which is newer than our initial timestamp, we activate "follow" mode. I also wonder if we really need to continue sorting the logs when we are in follow mode: I guess that printing them right away should be fine.

Am I making sense? :-)

Reply:
Yea definitely make sense.

> There is no case where we would get out of the outer loop because of the timeout, is there?
Exactly, the issue is only for the inner loop where we batch read logs.

> I'm not sure what the solution could be. We should activate the 25ms timer (for the follow mode) and send the first batch of logs only after having read the backlogs. Maybe: before the first iteration, we get the current time; then, once we read a log line which is newer than our initial timestamp, we activate "follow" mode. I also wonder if we really need to continue sorting the logs when we are in follow mode: I guess that printing them right away should be fine.

There is an obvious caveat here, what if the log never sends anything newer than the initial timestamp? Then we just never activate follow mode (unless we rely on a timeout solution, but this would then be just for follow mode). 

Also while you might say that printing them right away could be the answer (it could), logs could easily come out of order timewise.

Reply:
I also just want to add that I agree this is a very annoying solution and isn't perfect as we can never ensure that read all the backlogs (espcially because we can never know when a backlog ends in follow mode) and get them sorted correctly like this. 

The root issue is that journalctl doesn't give us the logs for units if they are in a namespace, and this is only reason this entire PR is even necessary to begin with. Optimally we wouldn't even need to provide a namespace for journalctl when requesting logs for specific units. I mean it should be able to get logs for units even without a provided namespace. In my opinion this should be possible.

Reply:
Mmm... indeed this is something that journalctl should be able to do. Unfortunately, looking at the source code it does not seem that one can specify the `--namespace` option more than once, but could you please try if any of these work?

1. Use the `--file=` option to let journalctl directly open the journal files; according to the man page, this can be given multiple times and journalctl should be able to merge all logs properly.
2. See if a combination of `--namespace=*` and specifying multiple`-u` options work.

Reply:
mmh, did we miss this before, at least reading the man page, it seems `--namespace=*` plus -u (but we used -u before) should work for us? we should try, it would mean just keeping most of the old code